Duplicate Metric Error when Tracing Enabled on Both WebServer and gRPC Server

## Environment Details
* Helidon Version: 2.5.5
* Helidon SE
* JDK version: 17
* OS: Any

----------

## Problem Description
I am experiencing an issue when attempting to add `Tracing` to both the `WebServer` and `gRPC Server` in my _Helidon_ application, which is configured with `MetricsSupport`. When I try to configure tracing for both servers, I receive the following error. It is worth noting that this error does **not** occur if I only configure Tracing for one server with `.tracer(tracer)`.

```
Exception in thread "main" java.lang.RuntimeException: No reflection exceptions should be thrown unless there is a fundamental error in your code set up.
	at io.jaegertracing.internal.metrics.Metrics.createMetrics(Metrics.java:76)
	at io.jaegertracing.internal.metrics.Metrics.<init>(Metrics.java:32)
	at io.jaegertracing.internal.metrics.Metrics.<init>(Metrics.java:28)
	at io.jaegertracing.Configuration.getTracerBuilder(Configuration.java:229)
	at io.helidon.tracing.jaeger.JaegerTracerBuilder.build(JaegerTracerBuilder.java:456)
	at io.helidon.tracing.jaeger.JaegerTracerBuilder.build(JaegerTracerBuilder.java:153)
	at io.helidon.common.Builder.get(Builder.java:41)
	at io.helidon.grpc.server.GrpcServerConfiguration$Builder.tracer(GrpcServerConfiguration.java:240)
	at com.example.tracing.Main.startServer(Main.java:71)
	at com.example.tracing.Main.main(Main.java:21)
Caused by: java.lang.IllegalArgumentException: Attempt to reregister the existing metric MetricID{name='jaeger_tracer_reporter_queue_length', tags=[]}
	at io.helidon.metrics.api.AbstractRegistry.enforceMetricUniqueness(AbstractRegistry.java:713)
	at io.helidon.metrics.api.AbstractRegistry.enforceMetricUniqueness(AbstractRegistry.java:708)
	at io.helidon.metrics.api.AbstractRegistry.registerUniqueMetric(AbstractRegistry.java:668)
	at io.helidon.metrics.api.AbstractRegistry.register(AbstractRegistry.java:118)
	at io.helidon.metrics.jaeger.HelidonJaegerMetricsFactory$3.<init>(HelidonJaegerMetricsFactory.java:88)
	at io.helidon.metrics.jaeger.HelidonJaegerMetricsFactory.createGauge(HelidonJaegerMetricsFactory.java:82)
	at io.jaegertracing.internal.metrics.Metrics.createMetrics(Metrics.java:66)
	... 9 more
```


## Steps to reproduce
This error can be reproduced with the following _Java_ code:
```java
    LogConfig.configureRuntime();

    var config = Config.create();

    var metrics = MetricsSupport.create();

    var tracer = TracerBuilder.create(config.get("tracing"));

    var simpleGreetService = new SimpleGreetService(config);
    var builder = Routing.builder()
        .register("/simple-greet", simpleGreetService);

    var health = HealthSupport.builder().addLiveness(HealthChecks.healthChecks()).build();
    var adminRouting = Routing.builder()
        .register(health)
        .register(metrics)
        .register(WebTracingConfig.create(config.get("tracing")))
        .build();

    // Web Server
    var server = WebServer.builder(builder)
        .tracer(tracer)
        .addNamedRouting("admin", adminRouting)
        .config(config.get("server"))
        .build();

    var webserver = server.start();
    webserver.thenAccept(ws -> ws.whenShutdown().thenRun(() -> System.out.println("WEB server is DOWN. Good bye!")))
        .exceptionallyAccept(t -> t.printStackTrace(System.err));

    // gRPC Server
    var tracingConfig = GrpcTracingConfig.builder()
        .withStreaming()
        .withVerbosity()
        .withTracedAttributes(ServerRequestAttribute.CALL_ATTRIBUTES,
            ServerRequestAttribute.HEADERS, ServerRequestAttribute.METHOD_NAME)
        .build();

    var grpcServerConfig =
        GrpcServerConfiguration.builder(config.get("grpc"))
            .tracer(tracer)
            .tracingConfig(tracingConfig)
            .build();

    var grpcRouting = GrpcRouting.builder()
        .register(new EchoService())
        .build();

    var grpcServer =
        GrpcServer.create(grpcServerConfig, grpcRouting);

    grpcServer.start().thenAccept(s -> s.whenShutdown().thenRun(() ->
        System.out.println("gRPC server is DOWN. Good bye!"))).exceptionally(t -> {
      System.out.println("Startup failed: " + t.getMessage());
      return null;
    });
```

--- a/metrics/jaeger/pom.xml
+++ b/metrics/jaeger/pom.xml
@@ -54,7 +54,22 @@
         </dependency>
         <dependency>
             <groupId>io.helidon.tracing</groupId>
+            <artifactId>helidon-tracing</artifactId>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>io.helidon.tracing</groupId>
             <artifactId>helidon-tracing-jaeger</artifactId>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>io.helidon.grpc</groupId>
+            <artifactId>helidon-grpc-server</artifactId>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>io.helidon.config</groupId>
+            <artifactId>helidon-config-yaml</artifactId>
             <scope>test</scope>
         </dependency>
         <dependency>

--- a/metrics/jaeger/src/main/java/io/helidon/metrics/jaeger/HelidonJaegerMetricsFactory.java
+++ b/metrics/jaeger/src/main/java/io/helidon/metrics/jaeger/HelidonJaegerMetricsFactory.java
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2021 Oracle and/or its affiliates.
+ * Copyright (c) 2021, 2023 Oracle and/or its affiliates.
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
@@ -15,7 +15,10 @@
  */
 package io.helidon.metrics.jaeger;
 
+import java.util.HashMap;
 import java.util.Map;
+import java.util.Optional;
+import java.util.SortedMap;
 import java.util.concurrent.TimeUnit;
 import java.util.function.BiFunction;
 
@@ -28,6 +31,7 @@ import io.jaegertracing.internal.metrics.Timer;
 import io.jaegertracing.spi.MetricsFactory;
 import org.eclipse.microprofile.metrics.Metadata;
 import org.eclipse.microprofile.metrics.Metric;
+import org.eclipse.microprofile.metrics.MetricID;
 import org.eclipse.microprofile.metrics.MetricRegistry;
 import org.eclipse.microprofile.metrics.MetricType;
 import org.eclipse.microprofile.metrics.MetricUnits;
@@ -40,6 +44,8 @@ public class HelidonJaegerMetricsFactory implements MetricsFactory {
 
     private final LazyValue<MetricRegistry> vendorRegistry = LazyValue.create(() -> RegistryFactory.getInstance()
             .getRegistry(MetricRegistry.Type.VENDOR));
+
+    private final Map<org.eclipse.microprofile.metrics.Gauge<?>, Gauge> jaegerGauges = new HashMap<>();
 
     @Override
     public Counter createCounter(String name, Map<String, String> jaegerTags) {
@@ -79,22 +85,28 @@ public class HelidonJaegerMetricsFactory implements MetricsFactory {
 
     @Override
     public Gauge createGauge(String name, Map<String, String> jaegerTags) {
-        return new Gauge() {
+        // Reusability does not apply to gauges, so explicitly check for a pre-existing gauge before registering a new one.
+        return existingGauge(name, jaegerTags)
+                .orElseGet(() ->
+                                   new Gauge() {
 
-            private final JaegerGauge gauge = new JaegerGauge();
+                                       private final JaegerGauge gauge = new JaegerGauge();
 
-            {
-                Metadata metadata = metadata(name, MetricType.GAUGE, MetricUnits.NONE);
-                vendorRegistry.get().register(metadata,
-                        gauge,
-                        convertTags(jaegerTags));
-            }
+                                       {
+                                           Metadata metadata = metadata(name, MetricType.GAUGE, MetricUnits.NONE);
+                                           // Register the new Helidon gauge and also add an entry in the
+                                           // Helidon-gauge-to-Jaeger-gauge map.
+                                           jaegerGauges.put(vendorRegistry.get().register(metadata,
+                                                                                          gauge,
+                                                                                          convertTags(jaegerTags)),
+                                                            this);
+                                       }
 
-            @Override
-            public void update(long amount) {
-                gauge.update(amount);
-            }
-        };
+                                       @Override
+                                       public void update(long amount) {
+                                           gauge.update(amount);
+                                       }
+                                   });
     }
 
     private org.eclipse.microprofile.metrics.Counter counter(Metadata metadata, Tag[] tags) {
@@ -152,5 +164,21 @@ public class HelidonJaegerMetricsFactory implements MetricsFactory {
                 .stream()
                 .map(e -> new Tag(e.getKey(), e.getValue()))
                 .toArray(Tag[]::new);
+    }
+
+    private Optional<Gauge> existingGauge(String name, Map<String, String> jaegerTags) {
+        SortedMap<MetricID, org.eclipse.microprofile.metrics.Gauge> existingGauges =
+                vendorRegistry.get()
+                        .getGauges((metricID, metric) -> metricID.getName().equals(name)
+                                && tagsMatch(metricID, jaegerTags));
+
+        return existingGauges.isEmpty()
+                ? Optional.empty()
+                : Optional.of(jaegerGauges.get(existingGauges.values().iterator().next()));
+    }
+
+    private static boolean tagsMatch(MetricID metricID, Map<String, String> jaegerTags) {
+        return jaegerTags == null && metricID.getTags().isEmpty()
+                || metricID.getTags().equals(jaegerTags);
     }
 }

--- a/metrics/jaeger/src/test/java/io/helidon/metrics/jaeger/TestMultipleJaegerClients.java
+++ b/metrics/jaeger/src/test/java/io/helidon/metrics/jaeger/TestMultipleJaegerClients.java
@@ -1,0 +1,88 @@
+/*
+ * Copyright (c) 2023 Oracle and/or its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.helidon.metrics.jaeger;
+
+import java.util.concurrent.TimeUnit;
+
+import io.helidon.config.Config;
+import io.helidon.grpc.server.GrpcRouting;
+import io.helidon.grpc.server.GrpcServer;
+import io.helidon.grpc.server.GrpcServerConfiguration;
+import io.helidon.grpc.server.GrpcTracingConfig;
+import io.helidon.metrics.api.RegistryFactory;
+import io.helidon.tracing.TracerBuilder;
+import io.helidon.webserver.Routing;
+import io.helidon.webserver.WebServer;
+
+import io.opentracing.Tracer;
+import org.eclipse.microprofile.metrics.MetricRegistry;
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+import static org.junit.jupiter.api.Assertions.fail;
+
+class TestMultipleJaegerClients {
+
+    @BeforeEach
+    @AfterEach
+    void clearVendorRegistry() {
+        RegistryFactory.getInstance().getRegistry(MetricRegistry.Type.VENDOR).removeMatching((metricId, Metric) -> true);
+    }
+
+    @Test
+    void checkMultipleJaegerClientsReuseMetric() {
+
+        Config config = Config.create();
+        Tracer tracer = TracerBuilder.create(config.get("tracing")).build();
+
+        WebServer webServer = WebServer.builder(createRouting(config))
+                .tracer(tracer)
+                .config(config.get("server"))
+                .build();
+
+        GrpcTracingConfig grpcTracingConfig = GrpcTracingConfig.create();
+        GrpcServerConfiguration grpcServerconfig = GrpcServerConfiguration.builder(config.get("grpc"))
+                .tracer(tracer)
+                .tracingConfig(grpcTracingConfig)
+                .build();
+
+        GrpcServer grpcServer = GrpcServer.create(grpcServerconfig, GrpcRouting.builder().build());
+
+        webServer.start()
+                .thenAccept(ws -> ws.whenShutdown().thenRun(() -> System.err.println("Web server has stopped.")))
+                .exceptionallyAccept(t -> fail("WebServer failed to start: " + t.getMessage()));
+
+        grpcServer.start()
+                .thenAccept(gs -> gs.whenShutdown().thenRun(() -> System.err.println("gRPC server has stopped.")))
+                .exceptionally(t -> {
+                    fail("gRPC server failed to start: " + t.getMessage());
+                    return null;
+                });
+
+        try {
+            TimeUnit.SECONDS.sleep(4);
+        } catch (InterruptedException e) {
+            fail("Error waiting for servers to warm up: ", e);
+        }
+    }
+
+    private Routing createRouting(Config config) {
+        return Routing.builder()
+                .get((req, resp) -> req.next())
+                .build();
+    }
+}

--- a/metrics/jaeger/src/test/resources/application.yaml
+++ b/metrics/jaeger/src/test/resources/application.yaml
@@ -1,0 +1,20 @@
+#
+# Copyright (c) 2023 Oracle and/or its affiliates.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+tracing:
+  service: twoClients
+
+app:
+  greeting: Hello

